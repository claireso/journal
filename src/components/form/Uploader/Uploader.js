import React, { useState, useRef } from 'react'
import PropTypes from 'prop-types'

import { create as createThumbnail } from '@services/thumbnails'
import { IconUpload } from '@components/Icons'

import * as S from './Uploader.styles'

const Uploader = (props) => {
  const [{ preview, error }, setState] = useState({ preview: props.preview })
  const input = useRef()

  const handleChange = async (event) => {
    const { accept, onChange } = props

    const files = event.target.files

    if (files.length === 0) return

    const file = files[0]

    if (!file) return

    if (!accept.includes(file.type)) {
      setState({
        preview: null,
        error: 'This file is not allowed'
      })
      input.current.value = ''
      return
    }

    try {
      const thumbnail = await createThumbnail(file)

      setState({ preview: thumbnail, error: null })

      onChange && onChange(thumbnail)
    } catch (err) {
      /* eslint-disable */
      console.error(err)
    }
  }

  return (
    <S.UploaderWrapper>
      {preview && (
        <S.UploaderPreview
          css={{
            background: props.backgroundPreview || '$secondary200'
          }}
        >
          <img src={preview} />
        </S.UploaderPreview>
      )}
      <S.UploaderContent>
        <IconUpload />
        <span>Upload new photo</span>
        <small>(only jpg and png)</small>
        <S.UploaderInput ref={input} type="file" name={props.name} onChange={handleChange} required={props.required} />
      </S.UploaderContent>
      {error && <S.UploaderError>{error}</S.UploaderError>}
    </S.UploaderWrapper>
  )
}

Uploader.propTypes = {
  name: PropTypes.string.isRequired,
  accept: PropTypes.array.isRequired,
  preview: PropTypes.string,
  required: PropTypes.bool,
  onChange: PropTypes.func
}

export default Uploader
