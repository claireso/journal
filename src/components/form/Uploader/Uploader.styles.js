import styled from 'styled-components'

export const UploaderWrapper = styled.div`
  border: 1px solid var(--border-color);
  border-radius: 0.4rem;
  box-shadow: 1px 1px 3px var(--box-shadow);
  overflow: hidden;
  position: relative;
`

export const UploaderPreview = styled.div`
  background: ${(props) => props.backgroundColor || 'var(--secondary-normal)'};
  padding: 2rem;

  img {
    display: block;
    max-width: 100%;
    max-height: 17rem;
    margin: 0 auto;
  }
`

export const UploaderContent = styled.div`
  padding: 2rem;

  svg {
    display: block;
    width: 2.4rem;
    height: 2.4rem;
    fill: var(--primary-normal);
    margin: 0 auto;
  }
  span {
    display: block;
    font-size: var(--font-size-normal);
    margin: 1rem 0 0;
    text-align: center;
  }
  small {
    color: var(--border-color-focus);
    display: block;
    font-size: var(--font-size-smaller);
    margin: 0.5rem 0 0;
    text-align: center;
  }
`

export const UploaderInput = styled.input`
  bottom: 0;
  cursor: pointer;
  left: 0;
  opacity: 0;
  position: absolute;
  right: 0;
  top: 0;
  width: 100%;
`

export const UploaderError = styled.div`
  color: var(--error-normal);
  font-size: var(--font-size-smaller);
  margin: 0.8rem 0 2rem;
  text-align: center;
`
