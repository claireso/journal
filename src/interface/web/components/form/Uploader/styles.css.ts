import { recipe } from '@vanilla-extract/recipes'
import { cmpStyles, tokens, p, gap, mx, my, radius } from '@web/theme'

export const wrapper = cmpStyles({
  overflow: 'hidden',
  position: 'relative',
  border: `1px solid ${tokens.colors.neutral.extralight}`,
  boxShadow: tokens.shadows.input,
  ...radius('size-1')
})

export const preview = recipe({
  base: cmpStyles({
    transition: 'opacity 120ms ease-out, background 150ms ease-out',
    ...p('size-5')
  }),
  variants: {
    processing: {
      true: cmpStyles({
        opacity: 0.5
      })
    }
  }
})

export const previewImage = cmpStyles({
  display: 'block',
  maxWidth: '100%',
  maxHeight: '17rem',
  ...mx('size-auto'),
  ...my('size-0')
})

export const content = cmpStyles({
  display: 'flex',
  alignItems: 'center',
  flexDirection: 'column',
  ...p('size-5'),
  ...gap('size-1')
})

export const input = cmpStyles({
  bottom: 0,
  cursor: 'pointer',
  left: 0,
  opacity: 0,
  position: 'absolute',
  right: 0,
  top: 0,
  width: '100%'
})
