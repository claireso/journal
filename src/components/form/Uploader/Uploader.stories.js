import React from 'react'
import { action } from '@storybook/addon-actions'

import Uploader from './index'

const params = {
  title: 'Form/Uploader',
  decorators: [(storyFn) => <div style={{ maxWidth: '500px' }}>{storyFn()}</div>]
}

export default params

export const Basic = () => <Uploader name="media" accept={['image/jpeg', 'image/png']} onChange={action('onChange')} />
