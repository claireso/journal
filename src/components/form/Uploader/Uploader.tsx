import React, { useState, useRef } from 'react'

import { create as createThumbnail } from '@services/thumbnails'
import { IconUpload } from '@components/Icons'

import * as S from './Uploader.styles'

interface UploaderProps {
  name: string
  accept: string[]
  preview?: string
  required?: boolean
  onChange?: (thumbnail: string) => void
  backgroundPreview?: string | null
}

interface State {
  preview?: string | null
  error?: string | null
}

const Uploader = (props: UploaderProps) => {
  const [{ preview, error }, setState] = useState<State>({ preview: props.preview })
  const input = useRef<HTMLInputElement>(null)

  const handleChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
    const { accept, onChange } = props

    const files = event.target.files

    if (!files || files.length === 0) return

    const file = files[0]

    if (!file) return

    if (!accept.includes(file.type)) {
      setState({
        preview: null,
        error: 'This file is not allowed'
      })
      if (input.current) {
        input.current.value = ''
      }
      return
    }

    try {
      const thumbnail = await createThumbnail(file)

      setState({ preview: thumbnail, error: null })

      onChange && onChange(thumbnail)
    } catch (err) {
      /* eslint-disable */
      console.error(err)
    }
  }

  return (
    <S.UploaderWrapper>
      {preview && (
        <S.UploaderPreview
          css={{
            background: props.backgroundPreview || '$secondary200'
          }}
        >
          <img src={preview} />
        </S.UploaderPreview>
      )}
      <S.UploaderContent>
        <IconUpload />
        <span>Upload new photo</span>
        <small>(only jpg and png)</small>
        <S.UploaderInput ref={input} type="file" name={props.name} onChange={handleChange} required={props.required} />
      </S.UploaderContent>
      {error && <S.UploaderError>{error}</S.UploaderError>}
    </S.UploaderWrapper>
  )
}

export default Uploader
