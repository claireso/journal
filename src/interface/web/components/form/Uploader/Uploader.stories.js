import React from 'react'
import { action } from '@storybook/addon-actions'

import Uploader from './index'

const params = {
  title: 'Form/Uploader',
  decorators: [(storyFn) => <div style={{ maxWidth: '500px' }}>{storyFn()}</div>],
  argTypes: { processing: { control: 'boolean' } },
  args: {
    processing: false
  }
}

export default params

export const Basic = (args) => (
  <Uploader
    name="media"
    accept={['image/jpeg', 'image/png']}
    onChange={action('onChange')}
    processing={args.processing}
  />
)
