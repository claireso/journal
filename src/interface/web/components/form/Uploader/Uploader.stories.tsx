import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import Uploader from './Uploader'

const meta: Meta<typeof Uploader> = {
  title: 'Form/Uploader',
  component: Uploader,
  argTypes: { processing: { control: 'boolean' } },
  args: {
    name: 'media',
    accept: ['image/jpeg', 'image/png'],
    processing: false,
    onChangeMedia: action('onChange'),
    onError: action('onError')
  }
}

export default meta
type Story = StoryObj<typeof Uploader>

export const Primary: Story = {}
