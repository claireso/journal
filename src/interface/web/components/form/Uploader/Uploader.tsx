import React, { useState, useRef, useCallback } from 'react'

import { create as createThumbnail } from '@web/services/thumbnails'

import { IconUpload } from '@web/components/Icons'
import Spinner from '@web/components/Spinner'

import * as S from './Uploader.styles'

interface UploaderProps {
  name: string
  accept: string[]
  processing: boolean
  preview?: string
  required?: boolean
  previewBackgroundColor?: string | null
  onError: (err: unknown) => void
  onChangeMedia: (file: File) => void
}

const Uploader = ({
  name,
  accept,
  processing,
  onChangeMedia,
  preview,
  required,
  previewBackgroundColor,
  onError
}: UploaderProps) => {
  const [currentPreview, setCurrentPreview] = useState<string | null>(preview ?? null)
  const [error, setError] = useState<string | null>(null)
  const refInput = useRef<HTMLInputElement>(null)

  const onChange = useCallback(
    async (event: React.ChangeEvent<HTMLInputElement>) => {
      const file = event.target.files?.[0]
      if (!file) {
        return
      }

      if (!accept.includes(file.type)) {
        setCurrentPreview(null)
        setError('This file type is not allowed')
        if (refInput.current) {
          refInput.current.value = ''
        }
        return
      }

      setError(null)

      try {
        const thumbnail = await createThumbnail(file)
        setCurrentPreview(thumbnail)
        onChangeMedia(file)
      } catch (err) {
        onError(err)
      }
    },
    [accept, onError, onChangeMedia]
  )

  return (
    <S.UploaderWrapper>
      {currentPreview && (
        <S.UploaderPreview
          css={{
            background: previewBackgroundColor || '$secondary200',
            opacity: processing ? 0.5 : 1
          }}
          data-testid="preview"
        >
          <img src={currentPreview} alt="" />
        </S.UploaderPreview>
      )}
      <S.UploaderContent>
        <>
          <S.UploaderIcon>{processing ? <Spinner /> : <IconUpload />}</S.UploaderIcon>
          <span>Upload new photo</span>
          <small>(only {accept.join(', ')})</small>
        </>

        <S.UploaderInput
          ref={refInput}
          type="file"
          name={name}
          onChange={onChange}
          required={required}
          accept={accept.join(',')}
        />
      </S.UploaderContent>
      {error && <S.UploaderError>{error}</S.UploaderError>}
    </S.UploaderWrapper>
  )
}

export default Uploader
