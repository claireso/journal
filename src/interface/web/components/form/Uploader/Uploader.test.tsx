import { render, fireEvent, waitFor } from '@testing-library/react'

import Uploader from './index'

describe('<Uploader />', () => {
  const renderComponent = (props = {}) =>
    render(
      <Uploader
        name="uploader"
        accept={['image/jpeg', 'image/png']}
        processing={false}
        onChangeMedia={jest.fn()}
        onError={jest.fn()}
        {...props}
      />
    )

  it('should render uploader', () => {
    const { asFragment } = renderComponent()

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render uploader with preview', () => {
    const { asFragment } = renderComponent({ preview: 'file.jpg' })

    expect(asFragment()).toMatchSnapshot()
  })

  it('should render error', () => {
    const { asFragment } = renderComponent()

    const input = document.querySelector('input[type="file"]')

    if (!input) {
      throw new Error('missing input in dom')
    }

    fireEvent.change(input, {
      target: {
        files: [new File(['(⌐□_□)'], 'mypdf.pdf', { type: 'application/pdf' })]
      }
    })

    expect(asFragment()).toMatchSnapshot()
  })

  it('should change file', async () => {
    const props = { onChangeMedia: jest.fn() }

    const { asFragment } = renderComponent(props)

    const input = document.querySelector('input[type="file"]')

    if (!input) {
      throw new Error('missing input in dom')
    }

    fireEvent.change(input, {
      target: {
        files: [new File(['(⌐□_□)'], 'mypicture.jpg', { type: 'image/jpeg' })]
      }
    })

    await waitFor(() => expect(props.onChangeMedia).toHaveBeenCalled())

    expect(props.onChangeMedia.mock.calls[0][0]).toBeInstanceOf(File)

    expect(asFragment()).toMatchSnapshot()
  })
})
