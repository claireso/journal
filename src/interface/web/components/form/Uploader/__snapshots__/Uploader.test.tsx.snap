// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Uploader /> should change file 1`] = `
<DocumentFragment>
  <div
    class="styles_layerStyle__xxgli80"
  >
    <div
      class="styles__xxgli83 styles_layerStyle__xxgli81"
    >
      <img
        alt=""
        class="styles_layerStyle__xxgli85"
        src="data:image/jpeg;base64,KOKMkOKWoV/ilqEp"
      />
    </div>
    <div>
      <div
        class="styles_layerStyle__xxgli86"
      >
        <svg
          aria-hidden="true"
          class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_lg__19o76jud styles_layerStyle__19o76ju4 styles_icon_variant_primary__19o76juf styles_layerStyle__19o76ju6"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 16.5v2.25A2.25 2.25 0 0 0 5.25 21h13.5A2.25 2.25 0 0 0 21 18.75V16.5m-13.5-9L12 3m0 0 4.5 4.5M12 3v13.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <span
          class="styles_text__kgzpwem styles_text_size_base__kgzpweq styles_layerStyle__kgzpwe3 styles_text_color_base__kgzpwev styles_layerStyle__kgzpwe8 styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          Upload new photo
        </span>
        <span
          class="styles_text__kgzpwem styles_text_size_sm__kgzpwep styles_layerStyle__kgzpwe2 styles_text_color_neutral__kgzpwex styles_layerStyle__kgzpwea styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          (only image/jpeg, image/png)
        </span>
      </div>
      <input
        accept="image/jpeg,image/png"
        class="styles_layerStyle__xxgli87"
        name="uploader"
        type="file"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Uploader /> should render error 1`] = `
<DocumentFragment>
  <div
    class="styles_layerStyle__xxgli80"
  >
    <div>
      <div
        class="styles_layerStyle__xxgli86"
      >
        <svg
          aria-hidden="true"
          class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_lg__19o76jud styles_layerStyle__19o76ju4 styles_icon_variant_primary__19o76juf styles_layerStyle__19o76ju6"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 16.5v2.25A2.25 2.25 0 0 0 5.25 21h13.5A2.25 2.25 0 0 0 21 18.75V16.5m-13.5-9L12 3m0 0 4.5 4.5M12 3v13.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <span
          class="styles_text__kgzpwem styles_text_size_base__kgzpweq styles_layerStyle__kgzpwe3 styles_text_color_base__kgzpwev styles_layerStyle__kgzpwe8 styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          Upload new photo
        </span>
        <span
          class="styles_text__kgzpwem styles_text_size_sm__kgzpwep styles_layerStyle__kgzpwe2 styles_text_color_neutral__kgzpwex styles_layerStyle__kgzpwea styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          (only image/jpeg, image/png)
        </span>
      </div>
      <input
        accept="image/jpeg,image/png"
        class="styles_layerStyle__xxgli87"
        name="uploader"
        type="file"
      />
    </div>
    <div
      class="styles__6gti4e6 styles_layerStyle__6gti4e0 styles_wrapper_status_error__6gti4e8 styles_layerStyle__6gti4e2"
    >
      This file type is not allowed
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Uploader /> should render uploader 1`] = `
<DocumentFragment>
  <div
    class="styles_layerStyle__xxgli80"
  >
    <div>
      <div
        class="styles_layerStyle__xxgli86"
      >
        <svg
          aria-hidden="true"
          class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_lg__19o76jud styles_layerStyle__19o76ju4 styles_icon_variant_primary__19o76juf styles_layerStyle__19o76ju6"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 16.5v2.25A2.25 2.25 0 0 0 5.25 21h13.5A2.25 2.25 0 0 0 21 18.75V16.5m-13.5-9L12 3m0 0 4.5 4.5M12 3v13.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <span
          class="styles_text__kgzpwem styles_text_size_base__kgzpweq styles_layerStyle__kgzpwe3 styles_text_color_base__kgzpwev styles_layerStyle__kgzpwe8 styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          Upload new photo
        </span>
        <span
          class="styles_text__kgzpwem styles_text_size_sm__kgzpwep styles_layerStyle__kgzpwe2 styles_text_color_neutral__kgzpwex styles_layerStyle__kgzpwea styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          (only image/jpeg, image/png)
        </span>
      </div>
      <input
        accept="image/jpeg,image/png"
        class="styles_layerStyle__xxgli87"
        name="uploader"
        type="file"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Uploader /> should render uploader with preview 1`] = `
<DocumentFragment>
  <div
    class="styles_layerStyle__xxgli80"
  >
    <div
      class="styles__xxgli83 styles_layerStyle__xxgli81"
    >
      <img
        alt=""
        class="styles_layerStyle__xxgli85"
        src="file.jpg"
      />
    </div>
    <div>
      <div
        class="styles_layerStyle__xxgli86"
      >
        <svg
          aria-hidden="true"
          class="styles__19o76ju9 styles_layerStyle__19o76ju0 styles_icon_size_lg__19o76jud styles_layerStyle__19o76ju4 styles_icon_variant_primary__19o76juf styles_layerStyle__19o76ju6"
          data-slot="icon"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 16.5v2.25A2.25 2.25 0 0 0 5.25 21h13.5A2.25 2.25 0 0 0 21 18.75V16.5m-13.5-9L12 3m0 0 4.5 4.5M12 3v13.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <span
          class="styles_text__kgzpwem styles_text_size_base__kgzpweq styles_layerStyle__kgzpwe3 styles_text_color_base__kgzpwev styles_layerStyle__kgzpwe8 styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          Upload new photo
        </span>
        <span
          class="styles_text__kgzpwem styles_text_size_sm__kgzpwep styles_layerStyle__kgzpwe2 styles_text_color_neutral__kgzpwex styles_layerStyle__kgzpwea styles_text_weight_normal__kgzpwe13 styles_layerStyle__kgzpweg"
        >
          (only image/jpeg, image/png)
        </span>
      </div>
      <input
        accept="image/jpeg,image/png"
        class="styles_layerStyle__xxgli87"
        name="uploader"
        type="file"
      />
    </div>
  </div>
</DocumentFragment>
`;
